### DISCUSSION TITLE
How to  implement custom shape plugin?

### DISCUSSION THREAD

**Discussion by kimren227** [mitsuba-renderer/mitsuba3#243] [Q&A]
Hi,

I wish to implement a custom shape plugin, i.e  Quadrics. What's the best way of doing that? Hope someone can point me to some starting documents or examples.

My Attemps so far:

- I've implemented the ray-shape intersection logic in dr.jit, and am able to render an image using examples in [Dr.Jit's First Steps in Python](https://drjit.readthedocs.io/en/latest/firststeps-py.html)
- I've read through [Mitsuba's custom plugin tutorial](https://mitsuba.readthedocs.io/en/latest/src/others/custom_plugin.html), however, I'm not sure what functions I need to implement and can't find a "register_shape()" function in the APIs.
- I've read the source code for "[shapes/sphere.cpp](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/shapes/sphere.cpp)", "[shapes/cylinder.cpp](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/shapes/cylinder.cpp)" etc, them seems all have implemented functions such as "surface_area", "sample_position", etc, are these necessary if I only need solid shape colors (Quadrics may have unbounded surface area)?

Many thanks in advance.

Daxuan

**Comment by Speierers** [comment]
Hi @kimren227 ,

Currently shape plugins can only be implemented in C++. In order to use your custom shape with OptiX, you will also need to provide a CUDA implementation of the `__intersection__myshape()` function. 

I would recommend you take a look at `src/shapes/sphere.cpp` and `src/shapes/optix/sphere.cuh`.

Regarding the implementation of the other methods, those are often necessary to use the shape as an area emitter. You can probably skip those if you are not planning on using it this way.

**Reply by njroussel** [reply]
Let me add this for completeness:

If you decide that you also want OptiX support for your shape, the current build process is a bit awkward. To point you in the right direction, take a look at `resources/ptx/Makefile`. This `Makefile` will generate `optix_rt.ptx` which has to be generated by hand **before** the usual `CMake` build.

**Reply by kimren227** [reply]
Thanks a lot for these information. I'll give it a shot!

**Reply by felixg13** [reply]
Hello @Speierers,

This thread is a few years old, and I was wondering whether the situation regarding the custom shape plugin has evolved. I noticed that a register_shape function now exists, but I can’t find any examples of a python custom shape plugin being created.

Thank you for any clarification.

**Reply by njroussel** [reply]
Hi @felixg13 

No, the setup is still the same and is unlikely to change.

As for `register_shape`, you'll find an example in `src/render/tests/test_shape.py`. It effectively is only really useful if you want to extend `mi.Mesh`, as the intersection routines are already defined.
